File: CSharp/StringHelpers/src/StringHelpers/RandomStringHelper.cs
using System.Buffers;
using System.Security.Cryptography;
using Microsoft.AspNetCore.WebUtilities;

namespace StringHelpers;

public static class RandomStringHelper
{
    private const string AlphabetLower = "abcdefghijklmnopqrstuvwxyz";
    private const string Digits = "0123456789";
    private const string Symbols = "`~!@#$%^&*()-_=+[{]}\\|;:'\",<.>/?";

    private static readonly char[] _availableCharactersAlphaNumericOnly;
    private static readonly char[] _availableCharactersPlusDashUnderscore;
    private static readonly char[] _availableCharacters;
    private static readonly char[] _uppercaseAlphaAndNumbers;

    static RandomStringHelper()
    {
        _availableCharactersAlphaNumericOnly = AlphabetLower
            .ToCharArray()
            .Concat(AlphabetLower.ToUpper())
            .Concat(Digits)
            .ToArray();

        _availableCharactersPlusDashUnderscore = _availableCharactersAlphaNumericOnly
            .Concat(new[] { '-', '_' })
            .ToArray();

        _availableCharacters = _availableCharactersAlphaNumericOnly
            .Concat(Symbols)
            .ToArray();

        _uppercaseAlphaAndNumbers = AlphabetLower
            .ToUpper()
            .ToCharArray()
            .Concat(Digits)
            .ToArray();
    }


    /// <summary>
    /// Generates a cryptographically-strong array of random bytes and return them as a base64 url-encoded string. 
    /// </summary>
    /// <param name="byteCount">The number of random bytes to generate.</param>
    /// <returns>The random bytes as a base64 url-encoded string.</returns>
    public static string GenerateRandomBase64UrlEncodedString(int byteCount)
    {
        if (byteCount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(byteCount), $"Invalid {nameof(byteCount)} value '{byteCount}'. It must be greater than 0.");
        }

        var randomStringBytes = ArrayPool<byte>.Shared.Rent(byteCount);
        try
        {
            GenerateRandomBytes(randomStringBytes, byteCount);

            return WebEncoders.Base64UrlEncode(randomStringBytes.AsSpan().Slice(0, byteCount));
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(randomStringBytes, clearArray: true);
        }
    }

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them as a base64-encoded string. 
    /// </summary>
    /// <param name="byteCount">The number of random bytes to generate.</param>
    /// <returns>The random bytes as a base64-encoded string.</returns>
    public static string GenerateRandomBase64EncodedString(int byteCount)
    {
        if (byteCount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(byteCount), $"Invalid {nameof(byteCount)} value '{byteCount}'. It must be greater than 0.");
        }

        var randomStringBytes = ArrayPool<byte>.Shared.Rent(byteCount);
        try
        {
            GenerateRandomBytes(randomStringBytes, byteCount);

            return Convert.ToBase64String(randomStringBytes.AsSpan().Slice(0, byteCount));
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(randomStringBytes, clearArray: true);
        }
    }

    /// <summary>
    /// Generates a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [A-Z0-9].
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [A-Z0-9].</returns>
    public static string GenerateUppercaseAlphanumericString(int length)
    {
        if (length <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(length), $"Invalid {nameof(length)} value '{length}'. It must be greater than 0.");
        }

        return InternalGenerateString(_uppercaseAlphaAndNumbers, length);
    }

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [a-zA-Z0-9-_].
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [a-zA-Z0-9-_].</returns>
    public static string GenerateAlphanumericString(int length, bool includeDashAndUnderscore = true)
    {
        if (length <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(length), $"Invalid {nameof(length)} value '{length}'. It must be greater than 0.");
        }

        var availableCharacters = includeDashAndUnderscore
            ? _availableCharactersPlusDashUnderscore
            : _availableCharactersAlphaNumericOnly;

        return InternalGenerateString(availableCharacters, length);
    }

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [a-zA-Z0-9-_], as well as the various symbol characters.
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [a-zA-Z0-9-_], as 
    /// well as the various symbol characters.</returns>
    public static string GenerateRandomString(int length)
    {
        if (length <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(length), $"Invalid {nameof(length)} value '{length}'. It must be greater than 0.");
        }

        return InternalGenerateString(_availableCharacters, length);
    }


    //
    // Private methods
    //

    private static string InternalGenerateString(char[] availableCharacters, int length)
    {
        // We need to control for modulo bias. Since we can't set the upper limit of bytes generated by
        //   RandomNumberGenerator, we do the next best thing and generate a uint32 of random data for 
        //   each character, and then take the modulo of those bytes. This effectively eliminates the
        //   modulo bias.
        // See: https://stackoverflow.com/a/1344255

        // Generate 4 bytes of randomness for each character.
        var multiplier = sizeof(uint);

        var randomStringBytes = ArrayPool<byte>.Shared.Rent(length * multiplier);
        try
        {
            GenerateRandomBytes(randomStringBytes, length * multiplier);

            return string.Create(
                length: length,
                state: new StringCreateArgs(AvailableCharacters: availableCharacters, RandomStringBytes: randomStringBytes, Length: length, Multiplier: multiplier),
                action: EncodeBytesAsCharacters
                );
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(randomStringBytes, clearArray: true);
        }
    }

    /// <summary>
    /// Fills a span with cryptographically strong random bytes.
    /// </summary>
    private static void GenerateRandomBytes(Span<byte> buffer, int length)
    {
        using var rng = RandomNumberGenerator.Create();

        // A pooled array is likely larger than the requested size. Only generate the requested number
        //   of bytes.
        rng.GetBytes(buffer.Slice(0, length));
    }

    private static void EncodeBytesAsCharacters(Span<char> destSpan, StringCreateArgs args)
    {
        // Remember, ArrayPool returns an array with a minimum size that is likely larger than what you
        //   request, so use the requested string length as the upper bound of the for loop, not the size
        //   of the random bytes array.
        var randomBytes = args.RandomStringBytes.AsSpan();

        for (var ixByte = 0; ixByte < args.Length; ixByte++)
        {
            // Get 4 bytes at a time.
            var randomNumber = BitConverter.ToUInt32(randomBytes.Slice(start: ixByte * args.Multiplier, length: args.Multiplier));

            // Mod by the number of available characters to index into the array.
            destSpan[ixByte] = args.AvailableCharacters[randomNumber % args.AvailableCharacters.Length];
        }
    }

    private record struct StringCreateArgs(char[] AvailableCharacters, byte[] RandomStringBytes, int Length, int Multiplier);
}

File: FSharp/StringHelpers/src/StringHelpers/RandomStringHelper.fs
﻿
namespace StringHelpers

/// Methods that generate random strings.
module RandomStringHelper =

    open System
    open System.Buffers
    open System.Runtime.CompilerServices
    open System.Security.Cryptography
    open Microsoft.AspNetCore.WebUtilities

    [<assembly:InternalsVisibleTo("StringHelpers.Tests")>]
    do
        ()


    //
    // Data
    //

    [<Literal>]
    let internal _alphabetLower = "abcdefghijklmnopqrstuvwxyz"
    [<Literal>]
    let internal _digits = "0123456789"
    [<Literal>]
    let private _symbols = "`~!@#$%^&*()-_=+[{]}\\|;:'\",<.>/?"

    let private _alphanumeric = Array.concat [|
        _alphabetLower |> Seq.toArray;
        _alphabetLower.ToUpper() |> Seq.toArray;
        _digits |> Seq.toArray;
        |]

    let internal _alphanumericPlusDashUnderscore = Array.concat [|
        _alphanumeric;
        [| '-'; '_' |];
        |]

    // Symbols includes dash and underscore.
    let private _alphanumericPlusSymbols = Array.concat [|
        _alphanumeric;
        _symbols |> Seq.toArray;
        |]

    let private _uppercaseAlphanumeric = Array.concat [|
        _alphabetLower.ToUpper() |> Seq.toArray;
        _digits |> Seq.toArray;
        |]


    //
    // Private functions
    //

    /// Generate a cryptographically-strong sequence of bytes.
    let private generateRandomBytes (buffer : Span<byte>) (length : int) =
        // A pooled array is likely larger than the requested size. Only generate the requested number
        //   of bytes.
        RandomNumberGenerator.Fill(buffer.Slice(0, length))


    //
    // Public functions
    //

    /// <summary>
    /// Generates a cryptographically-strong array of random bytes and return them as a base64 url-encoded string. 
    /// </summary>
    /// <remarks>The underlying <see cref="WebEncoders.Base64UrlEncode(ReadOnlySpan{byte})"/> removes any padding characters.</remarks>
    /// <param name="byteCount">The number of random bytes to generate.</param>
    /// <returns>The random bytes as a base64 url-encoded string.</returns>
    let generateRandomBase64UrlEncodedString (byteCount: int) =
        ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(byteCount, 0, (nameof(byteCount)))

        let randomBytes = ArrayPool<byte>.Shared.Rent(byteCount)
        try
            generateRandomBytes (randomBytes.AsSpan()) byteCount

            WebEncoders.Base64UrlEncode(randomBytes.AsSpan().Slice(0, byteCount));
        finally
            ArrayPool<byte>.Shared.Return(randomBytes, true)

    /// <summary>
    /// Decodes a string generated by the <see cref="generateRandomBase64UrlEncodedString(int)"/> method.
    /// </summary>
    /// <param name="base64UrlEncoded">A base64 url-encoded string to decode. Must not be null.</param>
    let decodeBase64UrlEncodedString (base64UrlEncoded : string) =
        ArgumentNullException.ThrowIfNull (nameof base64UrlEncoded)
        WebEncoders.Base64UrlDecode(base64UrlEncoded)

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them as a base64-encoded string. 
    /// </summary>
    /// <remarks>The underlying <see cref="Convert.ToBase64String(ReadOnlySpan{byte}, Base64FormattingOptions)"/> leaves
    /// any padding intact.</remarks>
    /// <param name="byteCount">The number of random bytes to generate.</param>
    /// <returns>The random bytes as a base64-encoded string.</returns>
    let generateRandomBase64EncodedString (byteCount: int) =
        ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(byteCount, 0, (nameof(byteCount)))

        let randomBytes = ArrayPool<byte>.Shared.Rent(byteCount)
        try
            generateRandomBytes (randomBytes.AsSpan()) byteCount

            Convert.ToBase64String(randomBytes.AsSpan().Slice(0, byteCount))
        finally
            ArrayPool<byte>.Shared.Return(randomBytes, true)

    /// <summary>
    /// Decodes a string generated by the <see cref="generateRandomBase64EncodedString(int)"/> method.
    /// </summary>
    /// <param name="base64Encoded">A base64 encoded string to decode. Must not be null.</param>
    let decodeBase64EncodedString (base64Encoded : string) =
        ArgumentNullException.ThrowIfNull (base64Encoded, (nameof base64Encoded))
        Convert.FromBase64String(base64Encoded)

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [A-Z0-9].
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [A-Z0-9].</returns>
    let generateUppercaseAlphanumericString (length: int) =
        ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(length, 0, (nameof length))

        RandomNumberGenerator.GetString(_uppercaseAlphanumeric, length)

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [a-zA-Z0-9].
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <param name="includeDashAndUnderscore">True to include '-' and '_' in the set of available characters;
    /// false to exclude them.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [a-zA-Z0-9].</returns>
    let generateAlphanumericString (length : int) (includeDashAndUnderscore : bool) =
        ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(length, 0, (nameof length))
        
        let characterSet = 
            match includeDashAndUnderscore with
            | true -> _alphanumericPlusDashUnderscore
            | false -> _alphanumeric

        RandomNumberGenerator.GetString(characterSet, length)

    /// <summary>
    /// Generate a cryptographically-strong array of random bytes and return them encoded as a string that
    /// can contain the characters in [a-zA-Z0-9-_], as well as the various symbol characters.
    /// </summary>
    /// <param name="length">The length of the random string to generate.</param>
    /// <returns>The random bytes encoded as a string that can contain the characters in [a-zA-Z0-9-_], as 
    /// well as the various symbol characters.</returns>
    let generateRandomString (length : int) =
        ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(length, 0, (nameof length))

        RandomNumberGenerator.GetString( _alphanumericPlusSymbols, length)


